inference-engine/src/gna_plugin/transformations/decompose_2d_conv.cpp

```diff
@@ -0,0 +1,672 @@
+// Copyright (C) 2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include <openvino/cc/ngraph/itt.hpp>
+
+#include "transformations/decompose_2d_conv.hpp"
+
+#include <memory>
+
+#include <ngraph/opsets/opset7.hpp>
+#include <ngraph/pattern/op/wrap_type.hpp>
+#include <ngraph/pattern/op/or.hpp>
+#include <ngraph/rt_info.hpp>
+#include <ngraph/pass/manager.hpp>
+#include <ie_common.h>
+#include "backend/gna_limitations.hpp"
+
+
+using namespace GNAPluginNS;
+
+NGRAPH_RTTI_DEFINITION(Decompose2DConv, "Decompose2DConv", 0);
+NGRAPH_RTTI_DEFINITION(Decompose2DConvTransposedWithBias, "Decompose2DConvTransposedWithBias", 0);
+NGRAPH_RTTI_DEFINITION(Decompose2DConvTransposedWithBiasAF, "Decompose2DConvTransposedWithBiasAF", 0);
+
+struct GraphData {
+    std::shared_ptr<ngraph::opset7::Transpose>leading_transpose;
+    std::shared_ptr<ngraph::opset7::Convolution>conv;
+    std::shared_ptr<ngraph::opset7::Transpose>trailing_transpose;
+    std::shared_ptr<ngraph::op::util::UnaryElementwiseArithmetic>af;
+    std::shared_ptr<ngraph::opset7::MaxPool>max_pool;
+    std::shared_ptr<ngraph::Node>last_op_in_sequence_for_replacement;
+    std::shared_ptr<ngraph::Node>bias_const;
+    size_t conv_count;
+    size_t pool_size_width;
+    size_t pool_stride_width;
+    // TODO: currently 2D max pool is not supported
+    //size_t pool_size_height;
+    //size_t pool_stride_height;
+};
+
+struct ConvParams {
+    size_t input_height;
+    size_t input_width;
+    size_t input_channel_count;
+    size_t filter_height;
+    size_t filter_width;
+    size_t filter_count;
+    size_t filter_channel_count;
+    size_t filter_dilation_height;
+    size_t filter_dilation_width;
+    size_t filter_stride_height;
+    size_t filter_stride_width;
+    size_t output_height;
+    size_t output_width;
+    size_t output_channel_count;
+    ngraph::element::Type element_type;
+};
+
+static bool VerifyAndGetConvParams(std::shared_ptr<ngraph::opset7::Convolution> conv, ConvParams& conv_params) {
+    const auto& input = conv->input_value(0);
+    const auto& filters = conv->input_value(1);
+
+    // We support only 2D conv batch 1
+    if (conv->get_dilations().size() != 2 ||
+        conv->get_strides().size() != 2 ||
```

Comment:
I don't think that it is necessary to check whether all parameters to consistent with 2D here. It is enough to check the rank of Convolution, another checks should be done during shape inference.

Reply:
Do you mean that verifying the convolution rank is enough to be sure that dilations and strides are given in two dimensions?

Reply:
@sirzabek otherwise convolution is not supposed to be valid